hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/util/RowDataAvroQueryContexts.java

```diff
@@ -38,17 +40,19 @@
  * Maintains auxiliary utilities for row data fields handling.
  */
 public class RowDataAvroQueryContexts {
-  private static final Map<Schema, RowDataQueryContext> QUERY_CONTEXT_MAP = new ConcurrentHashMap<>();
+  private static final Map<Pair<Schema, Boolean>, RowDataQueryContext> QUERY_CONTEXT_MAP = new ConcurrentHashMap<>();
 
   // BinaryRowWriter in RowDataSerializer are reused, and it's not thread-safe.
   private static final ThreadLocal<Map<Schema, RowDataSerializer>> ROWDATA_SERIALIZER_CACHE = ThreadLocal.withInitial(HashMap::new);
 
+  private static final Map<Triple<Schema, Schema, Map<String, String>>, RowProjection> ROW_PROJECTION_CACHE = new ConcurrentHashMap<>();
+
   public static RowDataQueryContext fromAvroSchema(Schema avroSchema) {
     return fromAvroSchema(avroSchema, true);
   }
 
   public static RowDataQueryContext fromAvroSchema(Schema avroSchema, boolean utcTimezone) {
-    return QUERY_CONTEXT_MAP.computeIfAbsent(avroSchema, k -> {
+    return QUERY_CONTEXT_MAP.computeIfAbsent(Pair.of(avroSchema, utcTimezone), k -> {
```

Comment:
I guess the `utcTimezone` should always be the same either for the write or read path?

Reply:
the cache is per JVM process, there exists case that the cache would be hit incorrectly, such as when flink session deployment mode is used, the cache maybe used among different jobs.